### PR TITLE
[ML] Early stopping in the line searches to compute initial regulariser values

### PR DESCRIPTION
**Pull request by tveasey**
This migrates to using BO for choosing points at which to evaluate the regularisers in the line searches we perform to find their initial values. Importantly, it also thresholds the minimum expected improvement to bother to continue with the line search to be at least 1% of the current test loss. This saves us *up to* 3 iterations in the loop (since we lower bound the minimum number of iterations we'll use) or around 40% of the cost of the line searches.

### CODE REVIEW

File: lib/api/unittest/CDataFrameAnalyzerFeatureImportanceTest.cc [hunk 1]
```diff
@@ -233,53 +221,54 @@ BOOST_FIXTURE_TEST_CASE(testRunBoostedTreeRegressionFeatureImportanceNoImportanc
...
+            double c2{result["row_results"]["results"]["ml"]["shap.c2"].GetDouble()};
+            double c3{result["row_results"]["results"]["ml"]["shap.c3"].GetDouble()};
+            double c4{result["row_results"]["results"]["ml"]["shap.c4"].GetDouble()};
+            double prediction{
+                result["row_results"]["results"]["ml"]["c5_prediction"].GetDouble()};
+            // c1 explain 97% of the prediction value, i.e. the difference from the prediction is less than 1%.
+            BOOST_REQUIRE_CLOSE(c1, prediction, 3.0);
+            BOOST_REQUIRE_SMALL(c2, 0.25);
+            BOOST_REQUIRE_SMALL(c3, 0.25);
+            BOOST_REQUIRE_SMALL(c4, 0.25);
```
[the remark this answers is not in the document]
Comment (tveasey) [reply lib/api/unittest/CDataFrameAnalyzerFeatureImportanceTest.cc]: These numbers are slightly unstable. The reason they were 0 before before is that none of these features were selected at all for training because of the threshold on MIC. Once they get selected there is scope for them to be non-zero. Since these are still so much smaller than c1, 0.25 vs ~100, I figured it was ok to just allow some leeway.

File: lib/maths/CBayesianOptimisation.cc [hunk 2]
```diff
@@ -16,10 +16,12 @@
 #include <maths/CLbfgs.h>
 #include <maths/CLinearAlgebraEigen.h>
 #include <maths/CLinearAlgebraShims.h>
+#include <maths/CMathsFuncs.h>
 #include <maths/CSampling.h>
 #include <maths/CTools.h>
 
 #include <boost/math/distributions/normal.hpp>
+#include <boost/optional/optional_io.hpp>
```
[the remark this answers is not in the document]
Comment (tveasey) [reply lib/maths/CBayesianOptimisation.cc]: This is needed to print out `boost::optional`.

[the remark this answers is not in the document]
Comment (valeriy42) [reply lib/maths/CBayesianOptimisation.cc]: I probably oversee something, by `boost/optional.hpp` is already include in the header. in the Line 177 I don't see any use of optional.

File: lib/api/unittest/CDataFrameAnalyzerFeatureImportanceTest.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (valeriy42) [reply lib/api/unittest/CDataFrameAnalyzerFeatureImportanceTest.cc]: :+1:

File: lib/maths/CBayesianOptimisation.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (valeriy42) [reply lib/maths/CBayesianOptimisation.cc]: The header `boost/optional.hpp` is include in the header file. Is it still required here? Line 177 doesn't use optional, afaics.

File: lib/api/unittest/CDataFrameAnalyzerFeatureImportanceTest.cc [hunk 3]
```diff
@@ -46,8 +49,6 @@ TDataFrameUPtr setupLinearRegressionData(const TStrVec& fieldNames,
         return core::CStringUtils::typeToStringPrecise(result[0], core::CIEEE754::E_DoublePrecision);
     };
 
-    auto frame = core::makeMainStorageDataFrame(weights.size() + 1).first;
```
[the remark this answers is not in the document]
Comment (tveasey) [reply lib/api/unittest/CDataFrameAnalyzerFeatureImportanceTest.cc]: No problem!

File: lib/maths/CBayesianOptimisation.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (tveasey) [reply lib/maths/CBayesianOptimisation.cc]: In this line I add `expectedImprovement` is type `boost::optional<double>`. This print state doesn't compile if I don't include `optional_io.hpp`
````
LOG_TRACE(<< "best = " << xmax.cwiseProduct(m_MaxBoundary - m_MinBoundary).transpose()
          << " EI(best) = " << expectedImprovement);

File: lib/maths/CBoostedTreeFactory.cc [hunk 4]
```diff
@@ -692,27 +716,96 @@ CBoostedTreeFactory::testLossLineSearch(core::CDataFrame& frame,
...
+        CBoostedTreeImpl::TNodeVecVec forest;
+        double testLoss;
+        std::tie(forest, testLoss) = m_TreeImpl->trainForest(
+            frame, m_TreeImpl->m_TrainingRowMasks[0], m_TreeImpl->m_TestingRowMasks[0],
+            m_TreeImpl->m_TrainingProgress, m_RecordMemoryUsage);
+        bopt.add(boptVector(regularizer), testLoss, 0.0);
+        minTestLoss.add(testLoss);
+        testLosses.emplace_back(regularizer, testLoss);
+    }
+    while (testLosses.size() > 0 && testLosses.size() < MAX_LINE_SEARCH_ITERATIONS) {
```
[the remark this answers is not in the document]
Comment (tveasey) [reply lib/maths/CBoostedTreeFactory.cc]: Yes. This can happen in the loop which searches for the best downsample factor, i.e. `applyRegularizer` can return false for all values in the loop
```
for (auto regularizer :
         {intervalLeftEnd, (2.0 * intervalLeftEnd + intervalRightEnd) / 3.0,
          (intervalLeftEnd + 2.0 * intervalRightEnd) / 3.0, intervalRightEnd}) {
  ...
```
In that case, we'll just use downsample factor of 1 and this actually SIGSEGVs when trying to compute maximum expected loss without this condition.

File: docs/CHANGELOG.asciidoc [hunk 5]
```diff
@@ -57,6 +57,8 @@ is no longer decreasing. (See {ml-pull}875[#875].)
 * Emit `prediction_field_name` in ml results using the type provided as
 `prediction_field_type` parameter. (See {ml-pull}877[#877].)
 * Improve performance updating quantile estimates. (See {ml-pull}881[#881].)
+* Migrate to BO for initial hyperparameter value line searches and stop early if the
```
[the remark this answers is not in the document]
Comment (tveasey) [reply docs/CHANGELOG.asciidoc]: See [f2fa5dd](https://github.com/elastic/ml-cpp/pull/903/commits/f2fa5dd3ec243b65cd70e6c1eee901df7847f802)

File: lib/maths/CBoostedTreeFactory.cc [hunk 6]
```diff
@@ -678,10 +701,11 @@ CBoostedTreeFactory::estimateTreeGainAndCurvature(core::CDataFrame& frame,
...
+                                        const TApplyRegularizer& applyRegularizer,
+                                        double intervalLeftEnd,
+                                        double intervalRightEnd,
                                         double returnedIntervalLeftEndOffset,
-                                        double returnedIntervalRightEndOffset,
-                                        double stepSize) const {
+                                        double returnedIntervalRightEndOffset) const {
 
     // This uses a quadratic approximation to the test loss function w.r.t.
     // the scaled regularization hyperparameter from which it estimates the
```
[the remark this answers is not in the document]
Comment (tveasey) [reply lib/maths/CBoostedTreeFactory.cc]: Good point. See [317c332](https://github.com/elastic/ml-cpp/pull/903/commits/317c332cb05e0f989b1bacaa1f015031b3871d89)

File: lib/maths/CBoostedTreeFactory.cc [hunk 7]
```diff
@@ -707,13 +707,19 @@ CBoostedTreeFactory::testLossLineSearch(core::CDataFrame& frame,
...
+    //   2. Fine tune, via Bayesian Optimisation targeting expected improvement,
+    //      and stop if the expected improvement small compared to the current
+    //      minimum test loss,
+    //   3. Calculate the parameter interval which gives the lowest test losses,
+    //   4. Fit an OLS quadratic approximation to the test losses in the interval
+    //      from step 3 and use it to estimate the best parameter value,
+    //   5. Compare the size of the residual errors w.r.t. to the OLS curve from
+    //      step 4 with its variation over the interval from step 3 and truncate
+    //      the returned interval if we can determine there is a low chance of
+    //      missing the best solution by doing so.
```
Comment (valeriy42) [inline lib/maths/CBoostedTreeFactory.cc]: nice! :+1:
